Make sure ItemsPresenter is updated before selection change.

grokys opened the pull request:
## What does the pull request do?

Having a separate subscription to `Items.CollectionChanged` in both `ItemsControl` and `ItemsPresenter` meant that the `ItemsPresenter` sometimes doesn't update itself in time for a selection change, resulting in #3148. Make `ItemsControl` trigger `ItemsPresenter.ItemsChanged` rather than both controls listening to the event separately.

## Fixed issues

Fixes #3148

MarchingCube commented:
Looks like this does not solve the issue after all, when checking I still get exception when typing in the search box.

grokys committed:
Validate min/max in MathUtilities.Clamp.

grokys committed:
Added failing test for #3148

grokys commented:
@MarchingCube yep, you're right. Ok, this time I've got a failing unit test, will work on a proper fix.

grokys committed:
Make sure ItemsPresenter is updated before selection change.

Having a separate subscription to `Items.CollectionChanged` in both `ItemsControl` and `ItemsPresenter` meant that the `ItemsPresenter` sometimes doesn't update itself in time for a selection change, resulting in #3148. Make `ItemsControl` trigger `ItemsPresenter.ItemsChanged` rather than both controls listening to the event separately.

grokys commented:
@MarchingCube I think I've fixed it this time - could you try again?

MarchingCube commented:
@grokys  Thanks! Repro case works great now 👍 . I will test tomorrow in my actual app, but since repro closely mimics our actual scenario I suspect it will work just fine.

grokys committed:
Added failing test.

`ItemsPresenter.Items` isn't correctly in sync with `ItemsControl.Items` when assigning new `Items`.

grokys committed:
Ensure Items are synchronised with Presenter.

Ensure that the `ItemsControl`'s presenter is notified of an `Items` change at the appropriate time.

MarchingCube commented:
I have found a new regression there:
```cs
[Fact]
        public void Logical_Parent_Is_Set_Correctly()
        {
            using (UnitTestApplication.Start(TestServices.StyledWindow))
            {
                ListBoxItem item = new ListBoxItem();

                var items = new List<ListBoxItem>
                {
                    item
                };

                var target = new ListBox
                {
                    Items = items,
                    VirtualizationMode = ItemVirtualizationMode.None
                };

                var root = new TestRoot(target)
                {
                    StylingParent = UnitTestApplication.Current
                };
                root.Measure(new Size(100, 100));
                root.Arrange(new Rect(0, 0, 100, 100));

                Assert.Equal(target, item.GetLogicalParent());
            }
        }
```

In current master it will pass, it fails after your changes. New logical parent is set to `VirtualizingStackPanel` and I think that it is not correct.

grokys committed:
Added another failing test.

Adapted from failing test described by @MarchingCube in PR feedback: https://github.com/AvaloniaUI/Avalonia/pull/3177#issuecomment-547515972

grokys committed:
Add to logical children before notifying presenter.

This makes sure that newly added control items have the correct logical parent.

grokys commented:
Modified that test slightly, and fixed the problem. Thanks for the repro!

MarchingCube committed:
Merge branch 'master' into fixes/3148-virtualization-out-of-range

MarchingCube reviewed (approved):
LGTM. New changes fix all issues that I was encountering. I don't see any more regressions but I will keep you posted.

grokys committed:
Merge branch 'master' into fixes/3148-virtualization-out-of-range
